### ISSUE TITLE
Errors: 'PosixPath' is not iterable and Expected to mark a variable ready only once.

### ISSUE THREAD

**Issue by wgb14** [k2-fsa/icefall#203]
2 errors when I ran librispeech example
```
conformer_ctc/train.py --full-libri false --world-size 2
```

```
Traceback (most recent call last):
  File "/home-2/gwang50/miniconda3/envs/k2/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/home-net/home-2/gwang50/code/icefall/egs/librispeech/ASR/conformer_ctc/train.py", line 604, in run
    if "lang_bpe" in params.lang_dir:
TypeError: argument of type 'PosixPath' is not iterable
```
```
Traceback (most recent call last):
  File "/home-2/gwang50/miniconda3/envs/k2/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/home-net/home-2/gwang50/code/icefall/egs/librispeech/ASR/conformer_ctc/train.py", line 680, in run
    scan_pessimistic_batches_for_oom(
  File "/home-net/home-2/gwang50/code/icefall/egs/librispeech/ASR/conformer_ctc/train.py", line 752, in scan_pessimistic_batches_for_oom
    loss.backward()
  File "/home-2/gwang50/miniconda3/envs/k2/lib/python3.8/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home-2/gwang50/miniconda3/envs/k2/lib/python3.8/site-packages/torch/autograd/__init__.py", line 154, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Expected to mark a variable ready only once. This error is caused by one of the following reasons: 1) Use of a module parameter outside the `forward` function. Please make sure mode
l parameters are not shared across multiple concurrent forward-backward passes. or try to use _set_static_graph() as a workaround if this module graph does not change during training loop.2) Reus
ed parameters in multiple reentrant backward passes. For example, if you use multiple `checkpoint` functions to wrap the same part of your model, it would result in the same set of parameters bee
n used by different reentrant backward passes multiple times, and hence marking a variable ready multiple times. DDP does not support such use cases in default. You can try to use _set_static_gra
ph() as a workaround if your module graph does not change over iterations.
Parameter at index 518 has been marked as ready twice. This means that multiple autograd engine  hooks have fired for this particular parameter during this iteration. You can set the environment
variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print parameter names for further debugging.
```
I can go on with training after the following fix, but idk if this break phone-based CTC training.
```diff
diff --git a/egs/librispeech/ASR/conformer_ctc/train.py b/egs/librispeech/ASR/conformer_ctc/train.py
index cb0bd5c..7498bcc 100755
--- a/egs/librispeech/ASR/conformer_ctc/train.py
+++ b/egs/librispeech/ASR/conformer_ctc/train.py
@@ -601,7 +601,7 @@ def run(rank, world_size, args):
     if torch.cuda.is_available():
         device = torch.device("cuda", rank)

-    if "lang_bpe" in params.lang_dir:
+    if "lang_bpe" in str(params.lang_dir):
         graph_compiler = BpeCtcTrainingGraphCompiler(
             params.lang_dir,
             device=device,
@@ -652,7 +652,7 @@ def run(rank, world_size, args):
     if world_size > 1:
         # Note: find_unused_parameters=True is needed in case we
         # want to set params.att_rate = 0 (i.e. att decoder is not trained)
-        model = DDP(model, device_ids=[rank], find_unused_parameters=True)
+        model = DDP(model, device_ids=[rank])

     optimizer = Noam(
         model.parameters(),
```

**Comment by csukuangfj** [comment]
Please also update the following line
https://github.com/k2-fsa/icefall/blob/27fa5f05d36323088d2cdd743cefcd9f05d266e1/egs/librispeech/ASR/conformer_ctc/train.py#L611

to

```python3
 elif "lang_phone" in str(params.lang_dir): 
```

---

> I can go on with training after the following fix, but idk if this break phone-based CTC training.

I don't think it will break the phone-based CTC training. Initially, `num_decoder_layers` was not zero and one had to use `find_unused_parameters=True`. Now all parameters in the phone-based model are used during training and that argument can be safely dropped.

**Comment by wgb14** [comment]
Thanks, Fangjun. I just tested this doesn't break phone-based training. Opened a PR https://github.com/k2-fsa/icefall/pull/211 to fix this.

**Comment by csukuangfj** [comment]
Closed via #211
